promremote/client.go

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package promremote
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/m3db/m3/src/query/ts"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/prometheus/prompb"
+)
+
+const (
+	// DefaultRemoteWrite is the default Prom remote write endpoint in m3coordinator.
+	DefaultRemoteWrite = "http://localhost:7201/api/v1/prom/remote/write"
+
+	defaulHTTPClientTimeout = time.Second * 30
+)
+
+// Tag are the metric tags.
+type Tag struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+
+// Timeseries are made of tags and a datapoint.
+type Timeseries struct {
+	Tags      []Tag
+	Datapoint ts.Datapoint
+}
+
+type client struct {
+	writeURL   string
+	httpClient *http.Client
+}
+
+type clientOptions struct {
+	remoteWriteURL    string
+	httpClientTimeout time.Duration
+}
+
+// M3Client is used to write timeseries data to m3coordinator.
+type M3Client interface {
+	// Write writes the Prom proto WriteRequest to the specified endpoint.
+	Write(*prompb.WriteRequest) error
+}
+
+// ClientOptions defines available methods.
+type ClientOptions interface {
```

Comment:
Are client options deserializable from e.g. yaml? I know we have this pattern a lot in the `m3` repo, but I personally find it rather cumbersome. How about something like what `zap` does? https://github.com/uber-go/zap/blob/master/config.go 

Advantages:

- Deserializable 
- Still has the possibility for defaults (see e.g. `NewProductionConfig`)

One other potential consideration:  I find the options pattern (see [Build](https://github.com/uber-go/zap/blob/master/config.go#L166)) to be pretty nice. It has the same niceties as the builder pattern, but you can define additional options in user code as well.

Reply:
+1 although i tend to prefer functional options in cases where struct initialization is cumbersome

Reply:
Yeah I would probably use a `Config` struct here, the M3 options pattern is better when a specific options is shared by thousands or millions of structs and they need to refer to a single pointer type.

Reply:
To avoid the struct initialization issue, you can probably just create a `func NewConfig(opts ...ConfigOption) Configuration {} ` constructor with functional options.

e.g.

```go
type Config struct {
  WriteURL string `yaml:"writeURL"`
  HTTPClientTimeout time.Duration
}

type ConfigOption func(*Config)

func NewConfig(opts ...ConfigOption) Config {
  var cfg Config
  for _, opt := range opts {
    opt(&cfg)
  }
  return cfg
}

func (c Config) Validate() error {
  // validate writeURL and httpclienttimeout
}

// Now the options
func WriteURLOption(writeURL string) ConfigOption {
  return func(c *Config) {
    c.WriteURL = writeURL
  }
}

func HTTPClientTimeoutOption(httpClientTimeout time.Duration) ConfigOption {
  return func(c *Config) {
    c.HTTPClientTimeout = httpClientTimeout
  }
}

// validating when you construct client
func NewClient(cfg Config) (PrometheusRemoteClient, error) {
  if err := cfg.Validate(); err != nil {
    return nil, err
  }

  return &client{...}, nil
}

// now using it
func foo() {
  cfg := NewConfig(
    WriteURLOption("http://foo:7201/bar"),
    HTTPClientTimeoutOption(30 * time.Second))

  client, err := NewClient(cfg)
  // check errr
}
```

Reply:
PS I would also allow users to pass in an HTTP client override if they have one already configured the way they like.

e.g.

```go
type Config struct {
  WriteURL string `yaml:"writeURL"`
  HTTPClientTimeout time.Duration `yaml:"httpClientTimeout"`
  HTTPClient *http.Client // if not-nil, use this instead of creating your own with the timeout as specified
}
```